File: 15. Check valid BST.py
"""
Task: Check valid BST
Time Complexity: O(n)
Space Complexity: O(h)
"""

class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

def is_bst_util(root,minv,maxv):
    if not root: return True
    if root.data<=minv or root.data>=maxv: return False
    return is_bst_util(root.left,minv,root.data) and is_bst_util(root.right,root.data,maxv)

if __name__=='__main__':
    root=Node(2); root.left=Node(1); root.right=Node(3)
    print('BST' if is_bst_util(root,float('-inf'), float('inf')) else 'Not BST')

File: 15. Check valid BST.cpp
#include <bits/stdc++.h>
using namespace std;
struct Node
{
    int data;
    Node* left;
    Node* right;
    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};
bool isBSTUtil(Node* root, long minv, long maxv)
{
    if (!root)
        return true;
    if (root->data <= minv || root->data >= maxv)
        return false;
    return isBSTUtil(root->left, minv, root->data) && isBSTUtil(root->right, root->data, maxv);
}
int main()
{
    Node* root = new Node(2);
    root->left = new Node(1);
    root->right = new Node(3);
    cout << (isBSTUtil(root, LONG_MIN, LONG_MAX) ? "BST" : "Not BST") << endl;
    return 0;
}
